apollo-federation/src/merger/merge_input.rs

```diff
@@ -0,0 +1,685 @@
+                } else {
+                    let mut present_subgraphs = Vec::new();
+                    let mut locations = Vec::new();
+
+                    // Extract nodes and create locations for fields that exist
+                    for (idx, field) in subgraph_fields.iter() {
+                        if let Some(field_component) = field {
+                            let _ = self
+                                .names
+                                .get(*idx)
+                                .map(|n| present_subgraphs.push(n.to_string()));
+                            if let Some(subgraph) = self.subgraphs.get(*idx) {
+                                let field_locations = subgraph
+                                    .schema()
+                                    .node_locations(&field_component.node)
+                                    .map(|loc| SubgraphLocation {
+                                        subgraph: subgraph.name.clone(),
+                                        range: loc,
+                                    });
+                                locations.extend(field_locations);
+                            }
+                        }
+                    }
+
+                    let present_subgraphs_str = present_subgraphs.into_iter().join(", ");
+                    let missing_subgraphs_str = missing_subgraphs.into_iter().join(", ");
+
+                    self.error_reporter.add_hint(CompositionHint {
+                        code: HintCode::InconsistentInputObjectField.code().to_string(),
+                        message: format!(
+                            "Input object field \"{}\" will not be added to \"{}\" in the supergraph as it does not appear in all subgraphs: it is defined in {} but not in {}",
+                            dest_field.field_name,
+                            dest.type_name,
+                            present_subgraphs_str,
+                            missing_subgraphs_str
+                        ),
+                        locations,
+                    });
+                }
+                // Note that we remove the element after the hint/error because we access the parent in the hint message.
+                dest_field.remove(&mut self.merged)?;
+            }
+        }
+
+        // We could be left with an input type with no fields, and that's invalid in GraphQL
+        let final_input_object = dest.get(self.merged.schema())?;
+        if final_input_object.fields.is_empty() {
+            let mut locations = Vec::new();
+            for (idx, input_type) in sources.iter() {
+                if let Some(input_component) = input_type {
+                    if let Some(subgraph) = self.subgraphs.get(*idx) {
+                        let type_locations =
+                            subgraph
+                                .schema()
+                                .node_locations(input_component)
+                                .map(|loc| SubgraphLocation {
```

Comment:
Can we just use `Merger::source_locations` method here? Or, we can refactor it to be more widely applicable.